File: LC/Medium/362.py
class HitCounter:

    def __init__(self):
        self.cache = [[0,i+1] for i in range(300)]
        
    def hit(self, timestamp: int) -> None:
        idx = (timestamp - 1)%300
        if self.cache[idx][1] == timestamp:
            self.cache[idx][0] += 1
        else:
            self.cache[idx][0] = 1            
            self.cache[idx][1] = timestamp   

    def getHits(self, timestamp: int) -> int:
        cnt = 0
        for x in self.cache:
            c,t = x[0],x[1]
            if c > 0 and timestamp - t < 300:
                cnt += c
        return cnt


# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)

File: LC/C++/Medium/362.cpp
class HitCounter {
private:
    std::vector<pair<int, int>> cache;

public:
    HitCounter() {
        cache.resize(300);
        for (size_t i = 0; i < 300; ++i) {
            cache[i] = {i + 1, 0};
        }
    }
    
    void hit(int timestamp) {
        int i = timestamp % 300;
        if (cache[i].first == timestamp) {
            cache[i].second++;
        } else {
            cache[i] = {timestamp, 1};
        }
    }
    
    int getHits(int timestamp) {
        int ans = 0;

        for (auto& [time, hits] : cache) {
            if (hits > 0 && timestamp - time < 300) {
                ans += hits;
            }
        }

        return ans;
    }
};

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter* obj = new HitCounter();
 * obj->hit(timestamp);
 * int param_2 = obj->getHits(timestamp);
 */